app/controllers/herbaria_controller.rb

```diff
@@ -0,0 +1,273 @@
+  def index
+    query = find_or_create_query(:Herbarium, by: params[:by])
+    show_selected_herbaria(query, id: params[:id].to_s, always_index: true)
+  end
+
+  def show
+    @canonical_url = herbarium_url(params[:id])
+    @herbarium = find_or_goto_index(Herbarium, params[:id])
+  end
+
+  # ---------- Actions to Display forms -- (new, edit, etc.) -------------------
+
+  def new
+    @herbarium = Herbarium.new
+  end
+
+  def edit
+    @herbarium = find_or_goto_index(Herbarium, params[:id])
+    return unless @herbarium
+    return unless make_sure_can_edit!
+
+    @herbarium.place_name         = @herbarium.location.try(&:name)
+    @herbarium.personal           = @herbarium.personal_user_id.present?
+    @herbarium.personal_user_name = @herbarium.personal_user.try(&:login)
+  end
+
+  # ---------- Actions to Modify data: (create, update, destroy, etc.) ---------
+
+  def create
+    @herbarium = Herbarium.new(herbarium_params)
+    normalize_parameters
+    return render(:new) unless validate_herbarium!
+
+    @herbarium.save
+    @herbarium.add_curator(@user) if @herbarium.personal_user
+    notify_admins_of_new_herbarium unless @herbarium.personal_user
+    redirect_to_create_location_or_referrer_or_show_location
+  end
+
+  def update
+    return unless (@herbarium = find_or_goto_index(Herbarium, params[:id]))
+    return unless make_sure_can_edit!
+
+    @herbarium.attributes = herbarium_params
+    normalize_parameters
+    return unless validate_herbarium!
+
+    @herbarium.save
+    redirect_to_create_location_or_referrer_or_show_location
+  end
+
+  def destroy
+    return unless (@herbarium = find_or_goto_index(Herbarium, params[:id]))
+
+    if user_can_destroy_herbarium?
+      @herbarium.destroy
+      redirect_to_referrer ||
+        redirect_with_query(herbaria_path(id: @herbarium.try(&:id)))
+    else
+      flash_error(:permission_denied.t)
+      redirect_to_referrer || redirect_with_query(herbarium_path(@herbarium))
+    end
+  end
+
+  # ========== Non=standard REST Actions =======================================
+
+  ##############################################################################
+
+  private
+
+  include Herbaria::SharedPrivateMethods
+
+  def make_sure_can_edit!
+    return true if in_admin_mode? || @herbarium.can_edit?
+
+    flash_error(:permission_denied.t)
+    redirect_to_referrer || redirect_with_query(herbarium_path(@herbarium))
+    false
+  end
+
+  def normalize_parameters
+    [:name, :code, :email, :place_name, :mailing_address].each do |arg|
+      val = @herbarium.send(arg).to_s.strip_html.strip_squeeze
+      @herbarium.send("#{arg}=", val)
+    end
+    @herbarium.description = @herbarium.description.to_s.strip
+    @herbarium.code = "" if @herbarium.personal_user_id
+  end
+
+  def validate_herbarium!
+    validate_name! &&
+      validate_location! &&
+      validate_personal_herbarium! &&
+      validate_admin_personal_user!
+  end
```

Comment:
IMO. If possible, the validation should be done in the model. I'm not sure that's possible here, so I've made it into a separate story so as not to delay this PR.